HZ3-connector: not clear exception on start if HZ3 cluster is unavailable

HZ3-connector throws java.lang.reflect.InvocationTargetException if it tries to connect to unavailable HZ3 cluster. 
**STEPS TO REPRODUCE**
1. Make sure the HZ3 cluster is available. 
2. Start job to sink or source the data.
**RESULT**
Job fails with exception "java.lang.reflect.InvocationTargetException" followed by "java.lang.NullPointerException" exception
**ALTERNATIVE**
Configure HZ3-connector to connect to running HZ5 cluster and start it. As a result, the job also fails, you've got "java.lang.IllegalStateException: Unknown protocol: CB2" exception.
**EXPECTED RESULT**
The connector should throw a Jet exception with the clear reason for a fail ("HZ3 Cluster is not running" or "Version of a cluster is greater than supported")
**LOG**
```
Sep 09, 2021 2:53:58 PM com.hazelcast.internal.metrics.impl.MetricsConfigHelper
INFO: [LOCAL] [dev] [5.0-SNAPSHOT] Overridden metrics configuration with system property 'hazelcast.metrics.collection.frequency'='1' -> 'MetricsConfig.collectionFrequencySeconds'='1'
Sep 09, 2021 2:53:58 PM com.hazelcast.system.logo
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] 
	+       +  o    o     o     o---o o----o o      o---o     o     o----o o--o--o
	+ +   + +  |    |    / \       /  |      |     /         / \    |         |   
	+ + + + +  o----o   o   o     o   o----o |    o         o   o   o----o    |   
	+ +   + +  |    |  /     \   /    |      |     \       /     \       |    |   
	+       +  o    o o       o o---o o----o o----o o---o o       o o----o    o   
Sep 09, 2021 2:53:58 PM com.hazelcast.system
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
Sep 09, 2021 2:53:58 PM com.hazelcast.system
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Hazelcast Platform 5.0-SNAPSHOT (20210909 - 2fe6c5e) starting at [10.212.134.151]:5701
Sep 09, 2021 2:53:58 PM com.hazelcast.system
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Cluster name: dev
Sep 09, 2021 2:53:58 PM com.hazelcast.internal.metrics.impl.MetricsConfigHelper
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Collecting debug metrics and sending to diagnostics is enabled
Sep 09, 2021 2:53:59 PM com.hazelcast.instance.impl.Node
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Using Multicast discovery
Sep 09, 2021 2:53:59 PM com.hazelcast.cp.CPSubsystem
WARNING: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
Sep 09, 2021 2:53:59 PM com.hazelcast.jet.impl.JetServiceBackend
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Setting number of cooperative threads and default parallelism to 12
Sep 09, 2021 2:53:59 PM com.hazelcast.internal.diagnostics.Diagnostics
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
Sep 09, 2021 2:53:59 PM com.hazelcast.core.LifecycleService
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] [10.212.134.151]:5701 is STARTING
Sep 09, 2021 2:54:01 PM com.hazelcast.instance.impl.JetExtension
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Jet is enabled
Sep 09, 2021 2:54:01 PM com.hazelcast.internal.cluster.ClusterService
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [10.212.134.151]:5701 - d745a1f1-9898-4c20-b899-6cb637517861 this
]

Sep 09, 2021 2:54:01 PM com.hazelcast.core.LifecycleService
READ THE OBJECT
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] [10.212.134.151]:5701 is STARTED
Sep 09, 2021 2:54:01 PM com.hazelcast.internal.partition.impl.PartitionStateManager
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Initializing cluster partition table arrangement...
Sep 09, 2021 2:54:01 PM com.hazelcast.jet.impl.JobCoordinationService
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Starting job 06c6-d8f2-9340-0001 based on submit request
Sep 09, 2021 2:54:01 PM com.hazelcast.jet.impl.MasterJobContext
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Didn't find any snapshot to restore for job '06c6-d8f2-9340-0001', execution 06c6-d8f2-9341-0001
Sep 09, 2021 2:54:01 PM com.hazelcast.jet.impl.MasterJobContext
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Start executing job '06c6-d8f2-9340-0001', execution 06c6-d8f2-9341-0001, execution graph in DOT format:
digraph DAG {
	"remoteMapSource(h3-map)" [localParallelism=12];
	"mapSink(h5-map)" [localParallelism=1];
	"remoteMapSource(h3-map)" -> "mapSink(h5-map)" [label="partitioned", queueSize=1024];
}
HINT: You can use graphviz or http://viz-js.com to visualize the printed graph.
Sep 09, 2021 2:54:02 PM com.hazelcast.client.HazelcastClient
INFO: hz.client_1 [dev] [3.12.12] A non-empty group password is configured for the Hazelcast client. Starting with Hazelcast version 3.11, clients with the same group name, but with different group passwords (that do not use authentication) will be accepted to a cluster. The group password configuration will be removed completely in a future release.
Sep 09, 2021 2:54:02 PM com.hazelcast.client.spi.ClientInvocationService
INFO: hz.client_1 [dev] [3.12.12] Running with 2 response threads, dynamic=false
Sep 09, 2021 2:54:02 PM com.hazelcast.core.LifecycleService
INFO: hz.client_1 [dev] [3.12.12] HazelcastClient 3.12.12 (20210209 - 35096ec) is STARTING
Sep 09, 2021 2:54:02 PM com.hazelcast.core.LifecycleService
INFO: hz.client_1 [dev] [3.12.12] HazelcastClient 3.12.12 (20210209 - 35096ec) is STARTED
Sep 09, 2021 2:54:02 PM com.hazelcast.client.connection.nio.ClusterConnectorService
INFO: hz.client_1 [dev] [3.12.12] Trying to connect to cluster with name: dev
Sep 09, 2021 2:54:02 PM com.hazelcast.client.connection.nio.ClusterConnectorService
INFO: hz.client_1 [dev] [3.12.12] Trying to connect to [127.0.0.1]:7777 as owner member
Sep 09, 2021 2:54:02 PM com.hazelcast.client.connection.nio.ClusterConnectorService
WARNING: hz.client_1 [dev] [3.12.12] Exception during initial connection to [127.0.0.1]:7777: com.hazelcast.core.HazelcastException: java.net.SocketException: Connection refused to address /127.0.0.1:7777
Sep 09, 2021 2:54:02 PM com.hazelcast.client.connection.nio.ClusterConnectorService
WARNING: hz.client_1 [dev] [3.12.12] Unable to get live cluster connection, retry in 3000 ms, attempt 1 of 2.
Sep 09, 2021 2:54:05 PM com.hazelcast.client.connection.nio.ClusterConnectorService
INFO: hz.client_1 [dev] [3.12.12] Trying to connect to [127.0.0.1]:7777 as owner member
Sep 09, 2021 2:54:05 PM com.hazelcast.client.connection.nio.ClusterConnectorService
WARNING: hz.client_1 [dev] [3.12.12] Exception during initial connection to [127.0.0.1]:7777: com.hazelcast.core.HazelcastException: java.net.SocketException: Connection refused to address /127.0.0.1:7777
Sep 09, 2021 2:54:05 PM com.hazelcast.client.connection.nio.ClusterConnectorService
WARNING: hz.client_1 [dev] [3.12.12] Unable to get live cluster connection, attempt 2 of 2.
Sep 09, 2021 2:54:05 PM com.hazelcast.client.connection.nio.ClusterConnectorService
WARNING: hz.client_1 [dev] [3.12.12] Unable to connect to any address for cluster: dev. The following addresses were tried: [[127.0.0.1]:7777]
Sep 09, 2021 2:54:05 PM com.hazelcast.client.connection.nio.ClusterConnectorService
WARNING: hz.client_1 [dev] [3.12.12] Could not connect to any cluster, shutting down the client: Unable to connect to any cluster.
Sep 09, 2021 2:54:05 PM com.hazelcast.core.LifecycleService
INFO: hz.client_1 [dev] [3.12.12] HazelcastClient 3.12.12 (20210209 - 35096ec) is SHUTTING_DOWN
Sep 09, 2021 2:54:05 PM com.hazelcast.core.LifecycleService
INFO: hz.client_1 [dev] [3.12.12] HazelcastClient 3.12.12 (20210209 - 35096ec) is SHUTDOWN
Sep 09, 2021 2:54:05 PM com.hazelcast.jet.impl.operation.InitExecutionOperation
SEVERE: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] java.lang.reflect.InvocationTargetException
com.hazelcast.core.HazelcastException: java.lang.reflect.InvocationTargetException
	at com.hazelcast.connector.Hz3Util.createMapAdapter(Hz3Util.java:38)
	at com.hazelcast.connector.ReadMapOrCacheP$RemoteProcessorSupplier.init(ReadMapOrCacheP.java:205)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$bf6409b6$1(ExecutionPlan.java:333)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:516)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.initProcSuppliers(ExecutionPlan.java:332)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.initialize(ExecutionPlan.java:172)
	at com.hazelcast.jet.impl.execution.ExecutionContext.initialize(ExecutionContext.java:161)
	at com.hazelcast.jet.impl.JobExecutionService.lambda$initExecution$10(JobExecutionService.java:395)
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:531)
	at com.hazelcast.jet.impl.JobExecutionService.initExecution(JobExecutionService.java:395)
	at com.hazelcast.jet.impl.operation.InitExecutionOperation.doRun(InitExecutionOperation.java:96)
	at com.hazelcast.jet.impl.operation.AsyncOperation.run(AsyncOperation.java:54)
	at com.hazelcast.spi.impl.operationservice.Operation.call(Operation.java:196)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:273)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:249)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:214)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.run(OperationExecutorImpl.java:411)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.runOrExecute(OperationExecutorImpl.java:438)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeLocal(Invocation.java:600)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:579)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:540)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:240)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:59)
	at com.hazelcast.jet.impl.MasterContext.invokeOnParticipant(MasterContext.java:294)
	at com.hazelcast.jet.impl.MasterContext.invokeOnParticipants(MasterContext.java:277)
	at com.hazelcast.jet.impl.MasterJobContext.lambda$tryStartJob$2(MasterJobContext.java:238)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$52(JobCoordinationService.java:1226)
	at com.hazelcast.jet.impl.JobCoordinationService.submitToCoordinatorThread(JobCoordinationService.java:1235)
	at com.hazelcast.jet.impl.JobCoordinationService.submitToCoordinatorThread(JobCoordinationService.java:1225)
	at com.hazelcast.jet.impl.MasterJobContext.tryStartJob(MasterJobContext.java:197)
	at com.hazelcast.jet.impl.JobCoordinationService.tryStartJob(JobCoordinationService.java:1130)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitJob$1(JobCoordinationService.java:297)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$52(JobCoordinationService.java:1226)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$53(JobCoordinationService.java:1247)
	at com.hazelcast.internal.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:64)
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.hazelcast.connector.Hz3Util.createMapAdapter(Hz3Util.java:35)
	... 41 more
Caused by: java.lang.IllegalStateException: Unable to connect to any cluster.
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl.connectToClusterInternal(ClusterConnectorServiceImpl.java:214)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl.access$000(ClusterConnectorServiceImpl.java:62)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl$1.call(ClusterConnectorServiceImpl.java:279)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl$1.call(ClusterConnectorServiceImpl.java:275)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)

Sep 09, 2021 2:54:05 PM com.hazelcast.jet.impl.execution.ExecutionContext
SEVERE: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] job '06c6-d8f2-9340-0001', execution 06c6-d8f2-9341-0001 encountered an exception in ProcessorSupplier.close(), ignoring it
java.lang.NullPointerException
	at com.hazelcast.connector.ReadMapOrCacheP$RemoteProcessorSupplier.close(ReadMapOrCacheP.java:216)
	at com.hazelcast.jet.impl.execution.ExecutionContext.lambda$null$689bdc39$1(ExecutionContext.java:261)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:516)
	at com.hazelcast.jet.impl.execution.ExecutionContext.lambda$completeExecution$c2a8266$1(ExecutionContext.java:261)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:516)
	at com.hazelcast.jet.impl.execution.ExecutionContext.completeExecution(ExecutionContext.java:256)
	at com.hazelcast.jet.impl.JobExecutionService.lambda$completeExecution$559b0d71$1(JobExecutionService.java:528)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:516)
	at com.hazelcast.jet.impl.JobExecutionService.completeExecution(JobExecutionService.java:528)
	at com.hazelcast.jet.impl.JobExecutionService.terminateExecution0(JobExecutionService.java:716)
	at com.hazelcast.jet.impl.JobExecutionService.terminateExecution(JobExecutionService.java:707)
	at com.hazelcast.jet.impl.operation.TerminateExecutionOperation.run(TerminateExecutionOperation.java:58)
	at com.hazelcast.spi.impl.operationservice.Operation.call(Operation.java:196)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:273)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:249)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:214)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.run(OperationExecutorImpl.java:411)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.runOrExecute(OperationExecutorImpl.java:438)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeLocal(Invocation.java:600)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:579)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:540)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:240)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:59)
	at com.hazelcast.jet.impl.MasterContext.invokeOnParticipant(MasterContext.java:294)
	at com.hazelcast.jet.impl.MasterContext.invokeOnParticipants(MasterContext.java:277)
	at com.hazelcast.jet.impl.MasterJobContext.lambda$cancelExecutionInvocations$17(MasterJobContext.java:616)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)

Sep 09, 2021 2:54:05 PM com.hazelcast.jet.impl.MasterJobContext
SEVERE: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Execution of job '06c6-d8f2-9340-0001', execution 06c6-d8f2-9341-0001 failed
	Start time: 2021-09-09T14:54:01.800
	Duration: 00:00:03.507
	To see additional job metrics enable JobConfig.storeMetricsAfterJobCompletion
com.hazelcast.core.HazelcastException: java.lang.reflect.InvocationTargetException
	at com.hazelcast.connector.Hz3Util.createMapAdapter(Hz3Util.java:38)
	at com.hazelcast.connector.ReadMapOrCacheP$RemoteProcessorSupplier.init(ReadMapOrCacheP.java:205)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.lambda$initProcSuppliers$bf6409b6$1(ExecutionPlan.java:333)
	at com.hazelcast.jet.function.RunnableEx.run(RunnableEx.java:31)
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:516)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.initProcSuppliers(ExecutionPlan.java:332)
	at com.hazelcast.jet.impl.execution.init.ExecutionPlan.initialize(ExecutionPlan.java:172)
	at com.hazelcast.jet.impl.execution.ExecutionContext.initialize(ExecutionContext.java:161)
	at com.hazelcast.jet.impl.JobExecutionService.lambda$initExecution$10(JobExecutionService.java:395)
	at com.hazelcast.jet.impl.util.Util.doWithClassLoader(Util.java:531)
	at com.hazelcast.jet.impl.JobExecutionService.initExecution(JobExecutionService.java:395)
	at com.hazelcast.jet.impl.operation.InitExecutionOperation.doRun(InitExecutionOperation.java:96)
	at com.hazelcast.jet.impl.operation.AsyncOperation.run(AsyncOperation.java:54)
	at com.hazelcast.spi.impl.operationservice.Operation.call(Operation.java:196)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:273)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:249)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:214)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.run(OperationExecutorImpl.java:411)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.runOrExecute(OperationExecutorImpl.java:438)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeLocal(Invocation.java:600)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:579)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:540)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:240)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:59)
	at com.hazelcast.jet.impl.MasterContext.invokeOnParticipant(MasterContext.java:294)
	at com.hazelcast.jet.impl.MasterContext.invokeOnParticipants(MasterContext.java:277)
	at com.hazelcast.jet.impl.MasterJobContext.lambda$tryStartJob$2(MasterJobContext.java:238)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$52(JobCoordinationService.java:1226)
	at com.hazelcast.jet.impl.JobCoordinationService.submitToCoordinatorThread(JobCoordinationService.java:1235)
	at com.hazelcast.jet.impl.JobCoordinationService.submitToCoordinatorThread(JobCoordinationService.java:1225)
	at com.hazelcast.jet.impl.MasterJobContext.tryStartJob(MasterJobContext.java:197)
	at com.hazelcast.jet.impl.JobCoordinationService.tryStartJob(JobCoordinationService.java:1130)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitJob$1(JobCoordinationService.java:297)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$52(JobCoordinationService.java:1226)
	at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitToCoordinatorThread$53(JobCoordinationService.java:1247)
	at com.hazelcast.internal.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:64)
	at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.hazelcast.connector.Hz3Util.createMapAdapter(Hz3Util.java:35)
	... 41 more
Caused by: java.lang.IllegalStateException: Unable to connect to any cluster.
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl.connectToClusterInternal(ClusterConnectorServiceImpl.java:214)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl.access$000(ClusterConnectorServiceImpl.java:62)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl$1.call(ClusterConnectorServiceImpl.java:279)
	at com.hazelcast.client.connection.nio.ClusterConnectorServiceImpl$1.call(ClusterConnectorServiceImpl.java:275)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)

Sep 09, 2021 2:54:05 PM com.hazelcast.core.LifecycleService
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] [10.212.134.151]:5701 is SHUTTING_DOWN
Sep 09, 2021 2:54:05 PM com.hazelcast.internal.partition.impl.MigrationManager
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Shutdown request of Member [10.212.134.151]:5701 - d745a1f1-9898-4c20-b899-6cb637517861 this is handled
Sep 09, 2021 2:54:05 PM com.hazelcast.instance.impl.Node
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Shutting down multicast service...
Sep 09, 2021 2:54:05 PM com.hazelcast.instance.impl.Node
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Shutting down connection manager...
Sep 09, 2021 2:54:05 PM com.hazelcast.instance.impl.Node
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Shutting down node engine...
Sep 09, 2021 2:54:05 PM com.hazelcast.instance.impl.NodeExtension
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Destroying node NodeExtension.
Sep 09, 2021 2:54:05 PM com.hazelcast.instance.impl.Node
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] Hazelcast Shutdown is completed in 20 ms.
Sep 09, 2021 2:54:05 PM com.hazelcast.core.LifecycleService
INFO: [10.212.134.151]:5701 [dev] [5.0-SNAPSHOT] [10.212.134.151]:5701 is SHUTDOWN
```

--- a/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/main/java/com/hazelcast/connector/Hz3Sinks.java
+++ b/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/main/java/com/hazelcast/connector/Hz3Sinks.java
@@ -84,8 +84,8 @@ public final class Hz3Sinks {
      */
     @Beta
     @Nonnull
-    public static <K, V> Sink<Map.Entry<K, V>> map(@Nonnull String mapName, @Nonnull String clientXml) {
-        return map(mapName, Map.Entry::getKey, Map.Entry::getValue, clientXml);
+    public static <K, V> Sink<Map.Entry<K, V>> remoteMap(@Nonnull String mapName, @Nonnull String clientXml) {
+        return remoteMap(mapName, Map.Entry::getKey, Map.Entry::getValue, clientXml);
     }
 
     /**
@@ -105,13 +105,13 @@ public final class Hz3Sinks {
      */
     @Beta
     @Nonnull
-    public static <T, K, V> Sink<T> map(
+    public static <T, K, V> Sink<T> remoteMap(
             @Nonnull String mapName,
             @Nonnull FunctionEx<? super T, ? extends K> toKeyFn,
             @Nonnull FunctionEx<? super T, ? extends V> toValueFn,
             @Nonnull String clientXml) {
         return new SinkImpl<>("mapSink(" + mapName + ')',
-                writeMapP(mapName, toKeyFn, toValueFn, clientXml), toKeyFn);
+                writeRemoteMapP(mapName, toKeyFn, toValueFn, clientXml), toKeyFn);
     }
 
     /**
@@ -119,7 +119,7 @@ public final class Hz3Sinks {
      * {@link Sinks#map(String, FunctionEx, FunctionEx)}.
      */
     @Nonnull
-    static <T, K, V> ProcessorMetaSupplier writeMapP(
+    static <T, K, V> ProcessorMetaSupplier writeRemoteMapP(
             @Nonnull String mapName,
             @Nonnull FunctionEx<? super T, ? extends K> toKeyFn,
             @Nonnull FunctionEx<? super T, ? extends V> toValueFn,

--- a/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/main/java/com/hazelcast/connector/Hz3Util.java
+++ b/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/main/java/com/hazelcast/connector/Hz3Util.java
@@ -18,6 +18,7 @@ package com.hazelcast.connector;
 
 import com.hazelcast.connector.map.Hz3MapAdapter;
 import com.hazelcast.core.HazelcastException;
+import com.hazelcast.jet.JetException;
 
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
@@ -30,10 +31,12 @@ final class Hz3Util {
     static Hz3MapAdapter createMapAdapter(String clientXml) {
         try {
             Class<?> clazz = Thread.currentThread().getContextClassLoader()
-                    .loadClass("com.hazelcast.connector.map.impl.MapAdapterImpl");
+                                   .loadClass("com.hazelcast.connector.map.impl.MapAdapterImpl");
             Constructor<?> constructor = clazz.getDeclaredConstructor(String.class);
             return (Hz3MapAdapter) constructor.newInstance(clientXml);
-        } catch (InstantiationException | IllegalAccessException | InvocationTargetException
+        } catch (InvocationTargetException e) {
+            throw new JetException("Could not create instance of MapAdapterImpl", e.getCause());
+        } catch (InstantiationException | IllegalAccessException
                 | ClassNotFoundException | NoSuchMethodException e) {
             throw new HazelcastException(e);
         }

--- a/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/main/java/com/hazelcast/connector/ReadMapOrCacheP.java
+++ b/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/main/java/com/hazelcast/connector/ReadMapOrCacheP.java
@@ -213,7 +213,9 @@ final class ReadMapOrCacheP<F extends CompletableFuture, B, R> extends AbstractP
 
         @Override
         public void close(Throwable error) {
-            hz3MapAdapter.shutdown();
+            if (hz3MapAdapter != null) {
+                hz3MapAdapter.shutdown();
+            }
         }
 
         @Override

--- a/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/main/java/com/hazelcast/connector/WriteMapP.java
+++ b/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/main/java/com/hazelcast/connector/WriteMapP.java
@@ -30,7 +30,6 @@ import com.hazelcast.jet.core.ProcessorSupplier;
 import com.hazelcast.jet.core.Watermark;
 
 import javax.annotation.Nonnull;
-import javax.annotation.Nullable;
 import java.util.AbstractMap;
 import java.util.Collection;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -40,7 +39,6 @@ import java.util.function.Consumer;
 import java.util.stream.Stream;
 
 import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
-import static java.lang.Integer.max;
 import static java.util.stream.Collectors.toList;
 
 /**
@@ -171,9 +169,6 @@ final class WriteMapP<T, K, V> implements Processor {
         private final String mapName;
         private final FunctionEx<? super T, ? extends K> toKeyFn;
         private final FunctionEx<? super T, ? extends V> toValueFn;
-        private int maxParallelAsyncOps;
-
-        private transient Hz3MapAdapter hz3MapAdapter;
 
         Supplier(@Nonnull String clientXml, @Nonnull String mapName,
                         @Nonnull FunctionEx<? super T, ? extends K> toKeyFn,
@@ -183,12 +178,6 @@ final class WriteMapP<T, K, V> implements Processor {
             this.mapName = mapName;
             this.toKeyFn = toKeyFn;
             this.toValueFn = toValueFn;
-        }
-
-        @Override
-        public void init(@Nonnull Context context) throws Exception {
-            hz3MapAdapter = Hz3Util.createMapAdapter(clientXml);
-            maxParallelAsyncOps = max(1, MAX_PARALLELISM / context.localParallelism());
         }
 
         @Nonnull
@@ -202,9 +191,5 @@ final class WriteMapP<T, K, V> implements Processor {
                     .collect(toList());
         }
 
-        @Override
-        public void close(@Nullable Throwable error) throws Exception {
-            hz3MapAdapter.shutdown();
-        }
     }
 }

--- a/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/test/java/com/hazelcast/connector/BaseHz3Test.java
+++ b/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/test/java/com/hazelcast/connector/BaseHz3Test.java
@@ -68,6 +68,21 @@ public class BaseHz3Test extends HazelcastTestSupport {
             + "    </network>\n"
             + "</hazelcast-client>\n";
 
+    // Client configuration with port where there is no Hz 3 instance running
+    protected static final String HZ3_DOWN_CLIENT_CONFIG =
+            "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+            + "<hazelcast-client xmlns=\"http://www.hazelcast.com/schema/client-config\"\n"
+            + "                  xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\n"
+            + "                  xsi:schemaLocation=\"http://www.hazelcast.com/schema/client-config\n"
+            + "                  http://www.hazelcast.com/schema/client-config/hazelcast-client-config-3.12.xsd\">\n"
+            + "\n"
+            + "    <network>\n"
+            + "        <cluster-members>\n"
+            + "            <address>127.0.0.1:42000</address>\n"
+            + "        </cluster-members>\n"
+            + "    </network>\n"
+            + "</hazelcast-client>\n";
+
     protected HazelcastInstance hz3;
     protected Path customLibDir;
 

--- a/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/test/java/com/hazelcast/connector/Hz3EnrichmentTest.java
+++ b/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/test/java/com/hazelcast/connector/Hz3EnrichmentTest.java
@@ -22,6 +22,8 @@ import com.hazelcast.connector.map.Hz3MapAdapter;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.function.BiFunctionEx;
 import com.hazelcast.function.FunctionEx;
+import com.hazelcast.jet.JetException;
+import com.hazelcast.jet.Job;
 import com.hazelcast.jet.config.JobConfig;
 import com.hazelcast.jet.pipeline.BatchStage;
 import com.hazelcast.jet.pipeline.Pipeline;
@@ -40,6 +42,7 @@ import static com.hazelcast.connector.Hz3Enrichment.hz3ReplicatedMapServiceFacto
 import static com.hazelcast.connector.Hz3Enrichment.mapUsingIMap;
 import static com.hazelcast.connector.Hz3Enrichment.mapUsingIMapAsync;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 public class Hz3EnrichmentTest extends BaseHz3Test {
 
@@ -124,5 +127,31 @@ public class Hz3EnrichmentTest extends BaseHz3Test {
         JobConfig config = getJobConfig(mapStage.name());
         hz.getJet().newJob(p, config).join();
         assertThat(results).containsOnly("a", "b");
+    }
+
+    @Test
+    public void when_enrichFromInstanceDown_then_shouldThrowJetException() {
+        HazelcastInstance hz = createHazelcastInstance();
+        IList<String> results = hz.getList("result-list");
+        Pipeline p = Pipeline.create();
+
+        ServiceFactory<Hz3MapAdapter, AsyncMap<Integer, String>> hz3MapSF =
+                hz3MapServiceFactory("test-map", HZ3_DOWN_CLIENT_CONFIG);
+
+        BiFunctionEx<? super Map<Integer, String>, ? super Integer, String> mapFn =
+                mapUsingIMap(FunctionEx.identity(), (Integer i, String s) -> s);
+        BatchStage<String> mapStage = p.readFrom(TestSources.items(1, 2, 3))
+                                       .mapUsingService(
+                                               hz3MapSF,
+                                               mapFn
+                                       );
+        mapStage.writeTo(Sinks.list(results));
+
+        JobConfig config = getJobConfig(mapStage.name());
+        Job job = hz.getJet().newJob(p, config);
+
+        assertThatThrownBy(() -> job.join())
+                .hasStackTraceContaining(JetException.class.getName())
+                .hasStackTraceContaining("Unable to connect to any cluster");
     }
 }

--- a/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/test/java/com/hazelcast/connector/Hz3SinksTest.java
+++ b/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/test/java/com/hazelcast/connector/Hz3SinksTest.java
@@ -44,7 +44,7 @@ public class Hz3SinksTest extends BaseHz3Test {
                 new SimpleEntry<>(1, "a"),
                 new SimpleEntry<>(2, "b")
         );
-        Sink<Map.Entry<Integer, String>> sink = Hz3Sinks.map("test-map", HZ3_CLIENT_CONFIG);
+        Sink<Map.Entry<Integer, String>> sink = Hz3Sinks.remoteMap("test-map", HZ3_CLIENT_CONFIG);
         p.readFrom(source)
                 .writeTo(sink);
 
@@ -68,7 +68,7 @@ public class Hz3SinksTest extends BaseHz3Test {
                 new SimpleEntry<>(1, "a"),
                 new SimpleEntry<>(2, "b")
         );
-        Sink<Map.Entry<Integer, String>> sink = Hz3Sinks.map("test-map", HZ3_DOWN_CLIENT_CONFIG);
+        Sink<Map.Entry<Integer, String>> sink = Hz3Sinks.remoteMap("test-map", HZ3_DOWN_CLIENT_CONFIG);
         p.readFrom(source)
          .writeTo(sink);
 

--- a/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/test/java/com/hazelcast/connector/Hz3SinksTest.java
+++ b/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/test/java/com/hazelcast/connector/Hz3SinksTest.java
@@ -17,6 +17,8 @@
 package com.hazelcast.connector;
 
 import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.jet.JetException;
+import com.hazelcast.jet.Job;
 import com.hazelcast.jet.config.JobConfig;
 import com.hazelcast.jet.pipeline.BatchSource;
 import com.hazelcast.jet.pipeline.Pipeline;
@@ -28,6 +30,7 @@ import java.util.AbstractMap.SimpleEntry;
 import java.util.Map;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.assertj.core.api.Assertions.entry;
 
 public class Hz3SinksTest extends BaseHz3Test {
@@ -54,5 +57,26 @@ public class Hz3SinksTest extends BaseHz3Test {
                 entry(1, "a"),
                 entry(2, "b")
         );
+    }
+
+    @Test
+    public void when_writeToInstanceDown_then_shouldThrowJetException() {
+        HazelcastInstance hz = createHazelcastInstance();
+
+        Pipeline p = Pipeline.create();
+        BatchSource<SimpleEntry<Integer, String>> source = TestSources.items(
+                new SimpleEntry<>(1, "a"),
+                new SimpleEntry<>(2, "b")
+        );
+        Sink<Map.Entry<Integer, String>> sink = Hz3Sinks.map("test-map", HZ3_DOWN_CLIENT_CONFIG);
+        p.readFrom(source)
+         .writeTo(sink);
+
+        JobConfig config = getJobConfig(sink.name());
+        Job job = hz.getJet().newJob(p, config);
+
+        assertThatThrownBy(() -> job.join())
+                .hasStackTraceContaining(JetException.class.getName())
+                .hasStackTraceContaining("Unable to connect to any cluster");
     }
 }

--- a/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/test/java/com/hazelcast/connector/Hz3SourcesTest.java
+++ b/extensions/hazelcast-3-connector/hazelcast-3-connector-common/src/test/java/com/hazelcast/connector/Hz3SourcesTest.java
@@ -17,7 +17,10 @@
 package com.hazelcast.connector;
 
 import com.hazelcast.collection.IList;
+import com.hazelcast.config.Config;
+import com.hazelcast.core.Hazelcast;
 import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.jet.JetException;
 import com.hazelcast.jet.Job;
 import com.hazelcast.jet.config.JobConfig;
 import com.hazelcast.jet.pipeline.BatchSource;
@@ -30,6 +33,7 @@ import java.util.HashMap;
 import java.util.Map;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.assertj.core.api.Assertions.entry;
 
 public class Hz3SourcesTest extends BaseHz3Test {
@@ -103,4 +107,50 @@ public class Hz3SourcesTest extends BaseHz3Test {
         assertThat(result.entrySet()).isEqualTo(items.entrySet());
     }
 
+    @Test
+    public void when_readFromInstanceDown_then_shouldThrowJetException() {
+        HazelcastInstance hz = createHazelcastInstance();
+
+        Pipeline p = Pipeline.create();
+        BatchSource<Map.Entry<Integer, String>> source = Hz3Sources.remoteMap("test-map", HZ3_DOWN_CLIENT_CONFIG);
+        p.readFrom(source)
+         .map(Map.Entry::getValue)
+         .writeTo(Sinks.list("test-result"));
+
+        JobConfig config = getJobConfig(source.name());
+        Job job = hz.getJet().newJob(p, config);
+
+        assertThatThrownBy(() -> job.join())
+                .hasStackTraceContaining(JetException.class.getName())
+                .hasStackTraceContaining("Unable to connect to any cluster");
+    }
+
+    @Test
+    public void when_tryToReadFromHz5Cluster_then_shouldThrowJetException() {
+        Config hz5Config = new Config();
+        hz5Config.getNetworkConfig().setPort(15701);
+        HazelcastInstance hz5 = Hazelcast.newHazelcastInstance(hz5Config);
+
+        try {
+            HazelcastInstance hz = createHazelcastInstance();
+
+            Pipeline p = Pipeline.create();
+            BatchSource<Map.Entry<Integer, String>> source = Hz3Sources.remoteMap(
+                    "test-map",
+                    HZ3_CLIENT_CONFIG.replace("3210", "15701") // point to Hz 5 instance
+            );
+            p.readFrom(source)
+             .map(Map.Entry::getValue)
+             .writeTo(Sinks.list("test-result"));
+
+            JobConfig config = getJobConfig(source.name());
+            Job job = hz.getJet().newJob(p, config);
+
+            assertThatThrownBy(() -> job.join())
+                    .hasStackTraceContaining(JetException.class.getName())
+                    .hasStackTraceContaining("Unable to connect to any cluster");
+        } finally {
+            hz5.shutdown();
+        }
+    }
 }

